Rework "Home Pulumi Works" (#1567)
pulumi/docs: content/docs/reference/how.md

    weight: 6
---

Pulumi uses a desired state model for managing infrastructure. A Pulumi program is executed by a _language host_ to compute a desired state for a stack's infrastructure. The _deployment engine_ compares this desired state with the stack's current state and determines what resources need to be created, updated or deleted. The engine uses a set of _resource providers_ (such as [AWS]({{< relref "/docs/quickstart/aws" >}}), [Azure]({{< relref "/docs/quickstart/azure" >}}), [Kubernetes]({{< relref "/docs/quickstart/kubernetes" >}}), and so on.) in order to manage the individual resources.  As it operates, the engine updates the _state) of your infrastructure with information about all resources that have been provisioned as well as any pending operations.

The following diagram illustrates the interaction between these parts of the system:

<img src="/images/docs/reference/engine-block-diagram.png" alt="Pulumi engine and providers" width="600">

> Let's talk a little bit about each of these components and then see how they all fit together during an invocation of `pulumi up`

Reviewer: Yes, it doesn't belong in this doc but there are other typos and minor issues that are not introduced by this PR which I can address on a big editorial PR. We should just get this one merged once the original comments have been addressed. :)

## Language Hosts

The _language host_ is responsible for running a Pulumi program and setting up an environment where it can register resources with the _deployment engine_.  The language host is made up of two different pieces:

> 1. A language executor, which is a binary named `pulumi-language-<language-name>`, which Pulumi uses to the language your program is written in (e.g. Node or Python). This binary is distributed with the Pulumi CLI.

Author: Changed this to "which Pulumi uses to launch the runtime for the the language your program is written in", so it makes sense.

> 1. A language executor, which is a binary named `pulumi-language-<language-name>`, which Pulumi uses to the language your program is written in (e.g. Node or Python). This binary is distributed with the Pulumi CLI.

Reviewer 2: ```suggestion
1. A language executor, which is a binary named `pulumi-language-<language-name>`, that Pulumi uses to the language your program is written in (e.g. Node or Python). This binary is distributed with the Pulumi CLI.
```

> 2. A language runtime, which is responsible for preparing your program to be executed and observes its execution in order to detect resource registrations.  When a resource is _registered_ (via `new Resource()` in JavaScript or `Resource(...)` in Python) the language runtime communicates this request back to the _deployment engine_. The language runtime is distributed as a regular package, just like any other code you might depend on it your program.  For example, the Node runtime is contained in the `@pulumi/pulumi` package published to NPM and the Python runtime is contained in the `pulumi` package published to PyPI.

Author: I do like that more!

> 2. A language runtime, which is responsible for preparing your program to be executed and observes its execution in order to detect resource registrations.  When a resource is _registered_ (via `new Resource()` in JavaScript or `Resource(...)` in Python) the language runtime communicates this request back to the _deployment engine_. The language runtime is distributed as a regular package, just like any other code you might depend on it your program.  For example, the Node runtime is contained in the `@pulumi/pulumi` package published to NPM and the Python runtime is contained in the `pulumi` package published to PyPI.

Author: Sure!

## Deployment Engine

> The _deployment engine_ is responsible for computing the set of operations needed to drive the current state of your infrastructure into the desired state expressed by your program.  When a _resource registration_ is received from the language host, the engine consults the existing state to determine if that resource has been created before. If it has not, the engine uses a _resource provider_ in order to create it.  If it already exists, the engine works with the resource provider to determine what, if anything, has changed by comparing the old state of the resource with the new desired state of the resource as expressed by the program. If there are changes, the engine determines if it can _update_ the resource in place or if it must _replace_ it by _creating_ a new version and _deleting_ the old version. The decision depends on what properties of the resource are changing and the type of the resource itself.  When the language host communicates to the engine that it has completed execution of the Pulumi program, the engine looks for any existing resources that exist but it did not see a new resource registration for and schedules these resources for deletion.

Reviewer: ```suggestion
The _deployment engine_ is responsible for computing the set of operations needed to drive the current state of your infrastructure into the desired state expressed by your program.  When a _resource registration_ is received from the language host, the engine consults the existing state to determine if that resource has been created before. If it has not, the engine uses a _resource provider_ in order to create it.  If it already exists, the engine works with the resource provider to determine what, if anything, has changed by comparing the old state of the resource with the new desired state of the resource as expressed by the program. If there are changes, the engine determines if it can _update_ the resource in place or if it must _replace_ it by _creating_ a new version and _deleting_ the old version. The decision depends on what properties of the resource are changing and the type of the resource itself.  When the language host communicates to the engine that it has completed execution of the Pulumi program, the engine looks for any existing resources that it did not see a new resource registration for and schedules these resources for deletion.
```

> The deployment engine is embedded into the `pulumi` CLI itself.

Reviewer: ```suggestion
The deployment engine is embedded in the `pulumi` CLI itself.
```

## Resource Providers

A resource provider is made up of two different pieces:

> 1. A _resource plugin_, which is the binary used by the deployment engine to manage a resource. These plugins are stored in the _plugin cache_ (located in `~/.pulumi/plugins`) and can be manged with the `pulumi plugin` set of commands.

Reviewer: ```suggestion
1. A _resource plugin_, which is the binary used by the deployment engine to manage a resource. These plugins are stored in the _plugin cache_ (located in `~/.pulumi/plugins`) and can be managed using the `pulumi plugin` set of commands.
```

2. An _SDK_ which provides bindings for each type of resource the provider can manage.

> Like the language runtime itself, the SDKs are available as regular packages.  For example, the for Node there is an `@pulumi/aws` package published to NPM and a `pulumi_aws` package for Python published to PyPI.  When these packages are added to your project, they run `pulumi plugin install` behind the scenes in order to download the resource plugin from pulumi.com.

Reviewer: Can't suggest for the entire paragraph but please change to:

Like the language runtime itself, the SDKs are available as regular packages.  For example, there is a `@pulumi/aws` package for Node published to NPM and a `pulumi_aws` package for Python published to PyPI.  When these packages are added to your project, they run `pulumi plugin install` behind the scenes in order to download the resource plugin from pulumi.com.

> Like the language runtime itself, the SDKs are available as regular packages.  For example, the for Node there is an `@pulumi/aws` package published to NPM and a `pulumi_aws` package for Python published to PyPI.  When these packages are added to your project, they run `pulumi plugin install` behind the scenes in order to download the resource plugin from pulumi.com.

Author: Linked to both and added a link for `pulumi plugin install`

## Putting it all together

> Let's walk through a simple example, suppose we have the following Pulumi program, which creates two S3 buckets:

Reviewer: ```suggestion
Let's walk through a simple example. Suppose we have the following Pulumi program, which creates two S3 buckets:
```

```javascript
const mediaBucket = new aws.s3.Bucket("media-bucket");
const contentBucket = new aws.s3.Bucket("content-bucket");
```

Now, we run `pulumi stack init mystack`. Since `mystack` is a new stack, the "last deployed state" has no resources.

> Next, we run `pulumi up`. Since this program is written in JavaScript, the Pulumi CLI launches the Node language host and requests that it execute the program. When the first `aws.s3.Bucket` object is constructed, the language host sends as _resource registration_ request to the deployment engine and then continues executing the program. This is subtle, but important: _When the call to `new aws.s3.Bucket` returns, it does not mean that the actual S3 bucket has been created in AWS_, it just means the language host has expressed that this bucket is part of the desired state of your infrastructure.  The language host continues to execute your program concurrently with the engine processing this request.

Reviewer: Change "the language host sends as _resource registration_ " to "the language host sends a _resource registration_"

In this case, since the last deployed state has no resources, the engine determines that it needs to create the `media-bucket` resource. It uses the AWS resource plugin in order to create the resource and the AWS resource plugin uses the AWS SDK in order to go create it.  Note that the engine does not talk directly to AWS, instead it just asks the AWS Resource Plugin to create a Bucket. As new resource types are added, you can simply update the version of a resource provider to gain access to these new resources without having to update the CLI itself. When the operation to create this bucket is complete, the engine records information about the newly created resource in its state file.

As the engine was creating the `media-bucket` bucket, the language host continued to execute the Pulumi program. This caused another resource registration to be generated (for `content-bucket`). Since there is no dependency between these two buckets, the engine is able to process that request in parallel with the creation of `media-bucket`.

> After both operations have completed, the language host exists (as the program has finished running) and then engine and resource providers shutdown. The state for mystack now looks like the following:

Reviewer: ```suggestion
After both operations have completed, the language host exists as the program has finished running. Then the engine and resource providers shut down. The state for `mystack` now looks like the following:
```

```
stack mystack
   - aws.s3.Bucket "media-bucket653a4"
   - aws.s3.Bucket "content-bucket125ce"
```

> Note the extra suffixes on the end of these bucket names. This is due to a process called [auto-naming]({{< relref "/docs/reference/programming-model.md#autonaming" >}}), which Pulumi uses by default in order to allow you to deploy multiple copies of your infrastructure without creating name collisions for resources. This behavior can be disabled if desired.

Reviewer 3: > This behavior can be disabled if desired.

Is this documented anywhere on the site that we could link to from here?

Author: the section on auto-naming in the programming model page actually explains how to do this.

> Now, let's make a change to one of resources and run `pulumi up` again.  Since Pulumi operates on a desired state model, it will use the last deployed state to compute the minimal set of changes to needed to update your deployed infrastructure. For example, imagine that we wanted to make the S3 `media-bucket` publicly readable.  We change our program to express this new desired state:

Reviewer: Not edited in this PR (and I'm happy to make these changes if you like): 
> Now, let's make a change to one of our resources and run `pulumi up` again.

Author: Fixed!

> Now, let's make a change to one of resources and run `pulumi up` again.  Since Pulumi operates on a desired state model, it will use the last deployed state to compute the minimal set of changes to needed to update your deployed infrastructure. For example, imagine that we wanted to make the S3 `media-bucket` publicly readable.  We change our program to express this new desired state:

Reviewer: ```suggestion
Now, let's make a change to one of resources and run `pulumi up` again.  Since Pulumi operates on a desired state model, it will use the last deployed state to compute the minimal set of changes needed to update your deployed infrastructure. For example, imagine that we wanted to make the S3 `media-bucket` publicly readable.  We change our program to express this new desired state:

```

```javascript
const mediaBucket = new aws.s3.Bucket("media-bucket", {
    acl: "public-read",   // add acl
});
const contentBucket = new aws.s3.Bucket("content-bucket");
```

> When you run `pulumi preview` or `pulumi up`, the entire process starts over.  The language host starts running your program and the call to aws.s3.Bucket causes a new resource registration request to be sent to the engine. This time, however, our state already contains a resource named `media-bucket`, so engine asks the resource provider to compare the existing state from our previous run of `pulumi up` with the desired state expressed by the program. The process detects that the `acl` property has changed from `private` (the default value) to `public-read`. By again consulting the resource provider the engine determines that it is able to update this property without creating a new bucket, and so it tells the provider to update the acl property to `public-read`. When this operation completes, the current state is updated to reflect the change that had been made.

Reviewer 3: ```suggestion
When you run `pulumi preview` or `pulumi up`, the entire process starts over.  The language host starts running your program and the call to `aws.s3.Bucket` causes a new resource registration request to be sent to the engine. This time, however, our state already contains a resource named `media-bucket`, so engine asks the resource provider to compare the existing state from our previous run of `pulumi up` with the desired state expressed by the program. The process detects that the `acl` property has changed from `private` (the default value) to `public-read`. By again consulting the resource provider the engine determines that it is able to update this property without creating a new bucket, and so it tells the provider to update the acl property to `public-read`. When this operation completes, the current state is updated to reflect the change that had been made.
```

Nit: add backticks around `aws.s3.Bucket`.

The engine also receives a resource registration request for "content-bucket".  However, since there are no changes between the current state and the desired state, the engine does not need to make any changes to the resource.

Now, suppose we rename `content-bucket` to `app-bucket`.

```javascript
const mediaBucket = new aws.s3.Bucket("media-bucket", {
    acl: "public-read",
});
const contentBucket = new aws.s3.Bucket("app-bucket"); // renamed bucket
```

> This time, the engine will not need to make any changes to `media-bucket` since its desired state matches its actual state. However, when the resource request for `app-bucket` is processed, the engine sees there's no existing resource named `app-bucket` in the current state and so it must create a new S3 bucket.  Once that process is complete and the language host has shut down, the engine looks for any resources in the current state which it did not see resource registration for. In this case, since we removed the registration of `content-bucket` from our program, the engine calls the resource provider to delete the existing `content-bucket` bucket.

Reviewer: ```suggestion
This time, the engine will not need to make any changes to `media-bucket` since its desired state matches its actual state. However, when the resource request for `app-bucket` is processed, the engine sees there's no existing resource named `app-bucket` in the current state and so it must create a new S3 bucket.  Once that process is complete and the language host has shut down, the engine looks for any resources in the current state which it did not see a resource registration for. In this case, since we removed the registration of `content-bucket` from our program, the engine calls the resource provider to delete the existing `content-bucket` bucket.
```
